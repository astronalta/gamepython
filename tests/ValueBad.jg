

Vector < Value {
    x var Int
    y var Int
    z var Int
}

Params < Value {
    x var Int

    @init(x Int) {
        self.x = x
    }
}

NilAttr < Object {
    params Params
}

function(param Vector) {
    # Illegal, b/c params are passed by pointer and so modification of a value
    # type parameter is not allowed (side effects)
    param.x = 99
}

main() Int {
    # Test passing a value as a parameter to a func
    vec = Vector()
    function(vec)

    # Primitives and value types cannot be used in generics
    arr = Array[Vector](10)

    # Illegal decl of var w/o constructor
    params Params

    ret 0
}
