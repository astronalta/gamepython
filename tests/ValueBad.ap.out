ValueBad.ap:Modification of function param
ValueBad.ap:18:5: Uninitialized value type attribute 'params'
ValueBad.ap:33:11: Primitives and value types cannot be used in generics
ValueBad.ap:36:5: Uninitialized value type variable 'params'
